src/wallet/rpcwallet.cpp

```diff
@@ -2454,67 +2456,72 @@ Value zc_raw_receive(const json_spirit::Array& params, bool fHelp)
 
     LOCK(cs_main);
 
-    uint256 a_sk;
-    uint256 sk_enc;
+    SpendingKey k;
 
     {
         CDataStream ssData(ParseHexV(params[0], "zcsecretkey"), SER_NETWORK, PROTOCOL_VERSION);
         try {
-            ssData >> a_sk;
-            ssData >> sk_enc;
+            ssData >> k;
         } catch(const std::exception &) {
             throw runtime_error(
                 "zcsecretkey could not be decoded"
             );
         }
     }
 
-    libzerocash::PrivateAddress zcsecretkey(a_sk, sk_enc);
-    libzerocash::Address zcaddress(zcsecretkey);
-
     uint256 epk;
     unsigned char nonce;
     ZCNoteEncryption::Ciphertext ct;
+    uint256 h_sig;
 
     {
         CDataStream ssData(ParseHexV(params[1], "encrypted_bucket"), SER_NETWORK, PROTOCOL_VERSION);
         try {
             ssData >> nonce;
             ssData >> epk;
             ssData >> ct;
+            ssData >> h_sig;
         } catch(const std::exception &) {
             throw runtime_error(
                 "encrypted_bucket could not be decoded"
             );
         }
     }
 
-    libzerocash::Coin decrypted_bucket(ct, zcaddress, epk, nonce);
+    ZCNoteDecryption decryptor(k.viewing_key());
 
-    std::vector<unsigned char> commitment_v = decrypted_bucket.getCoinCommitment().getCommitmentValue();
-    uint256 commitment = uint256(commitment_v);
+    NotePlaintext npt = NotePlaintext::decrypt(
+        decryptor,
+        ct,
+        epk,
+        h_sig,
+        nonce
+    );
+    PaymentAddress payment_addr = k.address();
+    Note decrypted_note = npt.note(payment_addr);
 
     assert(pwalletMain != NULL);
-    libzcash::MerklePath path;
+    std::vector<boost::optional<ZCIncrementalWitness>> witnesses;
     uint256 anchor;
-    auto found_in_chain = pwalletMain->WitnessBucketCommitment(commitment, path, anchor);
-
-    CAmount value_of_bucket = decrypted_bucket.getValue();
+    uint256 commitment = decrypted_note.cm();
+    pwalletMain->WitnessBucketCommitment(
+        {commitment},
+        witnesses,
+        anchor
+    );
 
     CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
-    {
-        ss << decrypted_bucket.getValue();
-        ss << decrypted_bucket.getRho();
-        ss << decrypted_bucket.getR();
-    }
+    ss << npt;
```

Comment:
I am liking how the NotePlaintext abstraction simplifies things (not just here).